activesupport/lib/active_support/inflector/methods.rb

```diff
@@ -171,7 +179,14 @@ def upcase_first(string)
     #   downcase_first('I')                          # => "i"
     #   downcase_first('')                           # => ""
     def downcase_first(string)
-      string.length > 0 ? string[0].downcase.concat(string[1..-1]) : ""
+      case string.length
+      when 0
+        ""
+      when 1
+        string.downcase
+      else
+        string[0].downcase.concat(string[1, string.length])
```

Comment:
Would the following save us allocating a string for non-first characters?

```suggestion
        string = string.dup
        string[0] = string[0].downcase
```

Reply:
Or I guess maybe even

```suggestion
        first = string[0]
        if first.downcase!
          string = string.dup
          string[0] = first
        end
        string
```
to apply the same approach as above and skip the new string if there would be no changes.